subprojects/docs/src/docs/userguide/custom_gradle_types.adoc

```diff
@@ -0,0 +1,272 @@
+[[custom_gradle_types]]
+= Developing Custom Gradle Types
+
+There are several different kinds of "add-ons" to Gradle that you can develop, such as <<custom_plugins.adoc#custom_plugins,plugins>>, <<custom_tasks.adoc#custom_tasks,tasks>>,
+<<custom_plugins.adoc#sec:getting_input_from_the_build,project extensions>> or <<dependency_management_attribute_based_matching.adoc#implementing_artifact_transforms,artifact transforms>>,
+that are all implemented as classes and other types that can run on the JVM.
+This chapter discusses some of the features and concepts that are common to these types.
+You can use these features to help implement custom Gradle types and provide a consistent DSL for your users.
+
+This chapter applies to the following types:
+
+- Plugin types.
+- Task types.
+- Artifact transform parameters types.
+- Extension objects created using `ExtensionContainer.create()`, for example a project extension registered by a plugin.
+- Objects created using an `ObjectFactory.newInstance()`.
+- Elements of a `NamedDomainObjectContainer`.
+
+== Configuration using bean properties
+
+The custom Gradle types that you implement often hold some configuration that you want to make available to build scripts and other plugins.
+For example, a download task may have configuration that specifies the URL to download from and the file system location to write the result to.
+This configuration is represented as Java bean properties.
+
+Kotlin and Groovy provide conveniences for declaring Java bean properties, which make them good language choices to use to implement Gradle types.
+These conveniences are demonstrated in the samples below.
+
+[[managed_properties]]
+=== Managed properties
+
+Gradle provides some conveniences for implementing types with bean properties.
+Gradle can provide an implementation of a property. This is called a _managed property_, as Gradle takes care of managing the state of the property.
+A property may be _mutable_, meaning that it has both a getter method and setter method, or _read-only_, meaning that it has only a getter method.
+
+[NOTE]
+====
+Managed properties are currently an experimental <<feature_lifecycle.adoc#feature_lifecycle,incubating>> feature.
```

Comment:
I would leave out the `experimental` from the sentence. Incubating already implies experimental, right?

Reply:
true. I'll remove it